src/nnet3/nnet-simple-component.cc

```diff
@@ -668,12 +718,99 @@ void ClipGradientComponent::Backprop(const std::string &debug_info,
       in_deriv->ApplyCeiling(clipping_threshold_);
       in_deriv->ApplyFloor(-1 * clipping_threshold_);
     }
+
+    to_update->num_backproped_ += 1;
+    RepairGradients(debug_info, in_value, in_deriv, to_update);
   }
 }
 
+void ClipGradientComponent::RepairGradients(
+    const std::string &debug_info,
+    const CuMatrixBase<BaseFloat> &in_value,
+    CuMatrixBase<BaseFloat> *in_deriv, ClipGradientComponent *to_update) const {
+  KALDI_ASSERT(to_update != NULL);
+  BaseFloat default_clipped_proportion_threshold = 0.01;
+  BaseFloat default_clipping_to_value = 0.0;
+
+  // we use this 'repair_probability' (hardcoded for now) to limit
+  // this code to running on about half of the minibatches.
+  BaseFloat repair_probability = 0.5;
```

Comment:
On Thu, Jun 2, 2016 at 3:08 PM, Yiming Wang <notifications@github.com>
wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/803#discussion_r65599941>:
>
> >    }
> >  }
> >
> > +void ClipGradientComponent::RepairGradients(
> > +    const std::string &debug_info,
> > +    const CuMatrixBase<BaseFloat> &in_value,
> > +    CuMatrixBase<BaseFloat> *in_deriv, ClipGradientComponent *to_update) const {
> > +  KALDI_ASSERT(to_update != NULL);
> > +  BaseFloat default_clipped_proportion_threshold = 0.01;
> > +  BaseFloat default_clipping_to_value = 0.0;
> > +
> > +  // we use this 'repair_probability' (hardcoded for now) to limit
> > +  // this code to running on about half of the minibatches.
> > +  BaseFloat repair_probability = 0.5;
>
> In my BLSTM experiment, the self-repairs were called consistently at least
> between jobs within the same training iteration:
> train.3.1 1391 times out of 12180
> train.3.2 1200 times out of 12180
> train.3.3 1372 times out of 12180
> train.5.1 1639 times out of 12400
> train.5.2 1224 times out of 12180
> train.5.3 1344 times out of 12180
> train.6.3 600 times out of 12240
> train.8.2 432 times out of 12180
>

On your iterations 6 and 8 it was called a lot of times on just one of the
iterations, and never on the others.


> Perhaps there are two opposite forces when self-repair: one reducing the
> magnitude of gradient caused by reducing the magnitudes of parameter
> values, and the other one increasing the magnitude of gradient caused by
> reducing the inputs to nonlinearity on the path. It seems the former one
> dominates in this specific experiment.
>
You're forgetting that the extra part of the gradient, that we add, can
itself get backpropped through many frames.  So that would tend to increase
the magnitudes of the gradients directly.  But if it seems to be working
maybe we shouldn't worry too much.

What I was thinking of as a potential fix, was to only apply the
self-repair, say, every other minibatch, and only use the minibatches where
we're *not* applying self-repair to measure the gradients.
But this would be quite tricky to code, as it has to be done on the whole
minibatch to be effective (not on each component's backprop call, as there
can be many of those per minibatch).
Another idea is, when doing the self-repair, after we add the new term, to
scale down the entire matrix to have the same norm as before, so that the
magnitude doesn't change.

BTW, I have a theory about why it only happens in the backward version of
the BLSTM.  The iVector is estimated in a 'forward' causal way, so it
always sees the past.  But generally the iVectors in the future are more
accurate, as they've seen more info.  So maybe it's trying to propagate the
iVectors from the future to the past, resulting in 'forget' gates not being
active, and that sets the stage for instability.